### ISSUE TITLE
Polymorphic Associations

### ISSUE THREAD

**Issue by davidrivera** [balderdashy/waterline#484]
This is more of a question rather than an issue and the title speaks for itself. Are there any plans to support [polymorphic associations](http://guides.rubyonrails.org/association_basics.html#polymorphic-associations) in the future, much like how active record handles them?

**Comment by onceatime** [comment]
any news on this?

**Comment by dmarcelino** [comment]
@onceatime, this is not slated for [0.11](https://github.com/balderdashy/waterline/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.11) so it will probably take some time. Having said that, any help would be welcomed!

**Comment by onceatime** [comment]
ok will look into waterline and try to help. my situation is this i use ember as a frontend end sails as my api provider ( wont go back to rails again ). In ember i have pages and posts both of them are commentable.

do you see any workaround for this ( without blowing everything up with multiple models )....

in sails i have this for my page ( and also posts )
```
comments: {
  collection: 'comment',
  via: 'commentable'
}
```

and in comment ( right now )
```
commentable: {
  model: 'page'
}
```

what i can do is something like
```
commentablePage: {
  model: 'page'
},
commentablePost: {
  model: 'page'
}
```

but this feels ugly ;-)

any ideas?
in my mind a good syntax would be:
```
commentable: {
  model: ['page', 'post']
},
commentableType: {
  type: 'string'
}
```

**Comment by pixelig** [comment]
Any news here?

@onceatime i'm stuck with the same problem (also using ember as frontend). Did you find a solution or some kind of workaround?

**Comment by smith64fx** [comment]
Are there any news about this?

We are really happy if we could achieve this.

Our case is this.
We have a post.
A post has elements. Each element has an object, which can have any type.

For example:
Post ==> Element == object ==> Picture
        ==> Element == object ==> Video
        ==> Element == object ==> Movie
... and so on ...

So what we have to do is that we create models like picture, video etc. and make them inherit from the "object" model. So we can write:

Element.js
```
attributes:
    object: { 
        model: 'object'
    }
}
```

Object.js
```
attributes: {
    created_at: 'date',
    created_by::{
         model: 'user'
    }
}
```


Video.js
```
parentClass: 'object',
attributes: {
    length: 'float',
    title: 'string'
}
```

So now should be able to 

We were first trying to implement this functionality in our adapter, but the idea behind sailsjs is

**Comment by sailsbot** [comment]
Thanks for posting, @davidrivera.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by smith64fx** [comment]
We must find a solution here, this is an important feature!

**Comment by devinivy** [comment]
Since @particlebanana indicated that it's a valid feature request, I'm reopening until someone migrates it into the roadmap with a PR.

**Comment by sailsbot** [comment]
Thanks for posting, @davidrivera.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by didil** [comment]
any idea if this is going to be implemetend ?

**Comment by robbyoconnor** [comment]
This is a joke that this STILL hasn't been implemented

**Comment by devinivy** [comment]
@robbyoconnor waterline is open-source and I'm sure many would be pumped to have this, so feel free to spec, implement, and/or write tests for this feature.

**Comment by robbyoconnor** [comment]
@devinivy that wasn't supposed to come off so demanding -- just seems all development around sails is dead....

**Comment by kevinburkeshyp** [comment]
it came off as very demanding. "This is a joke" for a bunch of people who work on the project for free.



-kevin

**Comment by kevinburkeshyp** [comment]
It's really not acceptable to talk to the maintainers like that. You wouldn't talk like that in person so don't do it here.

-kevin

**Comment by robbyoconnor** [comment]
My apologies <3 - Just frustrated with the lack of maintenance...I do understand time constraints and free time lacking. :heart:

**Comment by kevinob11** [comment]
@robbyoconnor I similarly didn't feel like I had the experience to add this to waterline, but desperately wanted it and other features. Thankfully [Sails](https://github.com/balderdashy/sails) allows you to implement your own ORM if you want to! There are already hooks written for sequelize and I wrote one for bookshelf myself. Lots of options even if you don't feel you have the experience to contribute to waterline. Also the only way you are going to get the experience is jumping in and trying to help!

**Comment by robbyoconnor** [comment]
Wait, there are hooks for sequelize?

**Comment by robbyoconnor** [comment]
@kevinob11, [Found it](https://github.com/festo/sails-hook-sequelize)

**Comment by robbyoconnor** [comment]
This also exists: https://github.com/cesardeazevedo/sails-hook-sequelize-blueprints

**Comment by kevinob11** [comment]
@robbyoconnor and here is my attempt at (bookshelf)[https://github.com/praece/sails-bookshelf], of note it is super opinionated toward my current project, we may clean that up at the end of the project and share it more widely.

Also sorry if discussing alternatives in a project's issue queue isn't appropriate.

**Comment by emahuni** [comment]
any news on this?
I see strapi that uses waterline has polymorphic orm implemented. i don't understand why this doesnt have it. it seems like the guys at strapi worked a lot on the orm, there is a lot that was put on top. maybe we could borrow code from there, just my thought

**Comment by emahuni** [comment]
ok I have finished the first working draft and published to npm and github:
https://www.npmjs.com/package/sails-hook-polymorphic-orm
https://github.com/emahuni/sails-hook-polymorphic-orm

I will refine the README, was in a hurry when i wrote it. Tests should show you most of the usage.
